Delegated IPAM plugin

wedaly opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Add support to Cilium CNI for delegated IPAM plugins as described in the CNI spec: https://www.cni.dev/docs/spec/#delegated-plugin-protocol

I tested this in a [kind cluster](https://kind.sigs.k8s.io/) using the [reference CNI IPAM plugin host-local](https://github.com/containernetworking/plugins/tree/main/plugins/ipam/host-local) to assign IP addresses from `node.spec.PodCIDR`, with `enable-endpoint-health-checking=false` and `local-router-ipv4` set to prevent cilium-daemon from assigning itself an IP. Most of the `cilium connectivity test` suite passed in this configuration, except for the L7 policy tests (not sure why).

I also tested this in an AKS cluster configured to use Cilium with the reference "host-local" delegated plugin for IPAM. In this configuration, all of the Cilium connectivity tests passed.

Fixes: #issue-number

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1540,6 +1540,12 @@ func initEnv(cmd *cobra.Command) {
+	// Currently, cilium-agent tries to allocate itself IPs for health monitoring and cilium_host,
+	// so it will fail with delegated IPAM unless:
+	// (1) EnableHealthChecking is disabled, and (2) LocalRouterIPv4 is provided
+	// (which I think also implies native routing).
```
I think it's reasonable to check for something like IPAM == delegated + `option.Config.LocalRouterIPv4 == "" || option.Config.EnableHealthChecking` and fatal out. Probably the better place is in `c *DaemonConfig) Validate()` to do this sort of validation. We probably should think about the `LocalRouterIPv6` option as well, ie how do we make sure that this works in an expected way if either IPv4 or IPv6 is disabled.

It probably wouldn't hurt also for you to poke around in the other usages of these flags just to double-check that there aren't any other expectations for other flags being set when these flags are provided. Mostlikely it's just fine to assume that with delegated IPAM mode, the user must specify these two flags in this particular way though.

wedaly replied:
Makes sense, thanks!

wedaly replied:
added validation in `DaemonConfig.Validate()`

joestringer commented on pkg/ipam/ipam.go:
```diff
@@ -142,6 +142,15 @@ func NewIPAM(nodeAddressing types.NodeAddressing, c Configuration, owner Owner,
+		// TODO: this isn't safe, since cilium-agent might try to access these
+		// when IPV4Enabled() or IPV6Enabled() is true and cilium-agent is configured
+		// to assign itself a health endpoint IP and/or routerIP.
+		// Better to provide a no-op implementation instead?
```
An implementation that returns errors for any attempt to allocate IP addresses would be good IMO. This way at least it isn't a hard-crash.

wedaly replied:
added a no-op allocator

joestringer commented on pkg/ipam/option/option.go:
```diff
@@ -28,6 +28,10 @@ const (
+
+	// IPAMDelegatedPlugin is the value to select CNI delegated IPAM plugin mode.
+	// See https://www.cni.dev/docs/spec/#section-4-plugin-delegation
+	IPAMDelegatedPlugin = "delegated-plugin"
```
Worth noting that to use this mode, two pieces of configuration must be provided:

* For `cilium-agent`, `--ipam=delegated-plugin`
* For the CNI configuration, place the delegatee plugin name into the configuration

Given that Cilium's DaemonSet typically installs the CNI configuration file, it may make sense to update the Helm templates to allow the user to specify the IPAM delegation + other plugin in the helm options.

wedaly replied:
good call, will take a look at the helm chart as well

wedaly replied:
I decided not to tackle helm chart / installer changes in this PR, since it's already pretty large. For testing, we've been using the existing helm chart options to install a CNI configuration from a configmap.

joestringer replied:
Makes sense. It would be good to get that documented at some point so others can reproduce and test this setup.

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -389,25 +390,74 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
+		// again with a DEL command so it can release the IPs.
+		defer func() {
+			if err != nil {
+				cniInvoke.DelegateDel(context.TODO(), n.IPAM.Type, args.StdinData, nil)
```
Probably we should come up with a context early on in this whole function so that we can pass it down to each of these calls consistently and have a deadline that fits within the standard timeouts that e.g. Kubelet might expect. IIRC it's something like... kubelet has a 2 minute timeout so Cilium will try to accomplish everything within ~90s so that Cilium properly catches & implements any failures before returning.

wedaly replied:
I've left the `context.TODO()`s for now because I didn't want to make changes that could impact non-delegated-plugin codepaths. Interestingly, the bridge reference plugin also uses `context.TODO`, which makes me think it might be relying on the container runtime to kill the CNI process after some timeout.

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -389,25 +390,74 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
+		}
+
+		// The rest of the CNI code uses models.IPAMResponse, so convert the output of the
+		// IPAM plugin to that format. TODO: decide whether this is a reasonable approach.
```
Seems reasonable enough, this should allow the IPAM specific logic here to be relatively contained.

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -389,25 +390,74 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
+		for _, ipConfig := range ipamResult.IPs {
+			// TODO: there are two fields in ipamResult that we're ignoring here: Routes and Gateways.
+			// There doesn't seem to be a way to inject those into models.IPAMResponse, but maybe
+			// that's something we'd need to add?
```
Some recent context on this is #18975 (bug where we would unnecessarily churn kernel devices in the case where IPAM is limited). It may actually make sense to move more of the IPAM logic (ie the routes/gateways piece below) above the veth creation logic, so that the code is more self-contained to either invoke this new delegated mode or to figure out IPAM+routes+gateways all together in the standard codepaths of the existing case.

wedaly replied:
interesting, will take a look.

If I'm understanding the current implementation correctly, the routes are being derived from the host IP address? https://github.com/cilium/cilium/blob/2efbdd68a7a578874d1a9ecb884bfba4f76e0f0b/plugins/cilium-cni/cilium-cni.go#L237-L240 Would it make sense for Cilium CNI to include both that route and any returned by the delegated IPAM plugin?

joestringer replied:
These routes would be for inside the pod's namespace, so the main thing is to let it know about the Cilium gateway IP address which is terminated only on the L2 network between the pod and the other side of the veth pair (where Cilium implements an ARP responder). In most cases it should need just a default route with appropriate MTU, pointing towards that IP.

Then depending on the mode (eg `endpointRoutes`), Cilium may separately plumb routes on the host itself.

wedaly replied:
ah, that makes sense. Thanks!

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -630,6 +692,13 @@ func cmdDel(args *skel.CmdArgs) error {
+	// TODO: following the example of the bridge reference plugin here.
+	// If Netns is empty, call the delegated IPAM plugin DEL to release any IPs,
+	// but don't need to do anything else afterwards.
+	if conf.IpamMode == ipamOption.IPAMDelegatedPlugin && args.Netns == "" {
```
I'm not sure I quite understand this case. When would `args.Netns` be empty?

wedaly replied:
I cargo-culted this from the bridge reference plugin. Will check the commit history to see if there's any more context about why they added this.

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -389,25 +390,74 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
+				releaseIP(c, ipam.Address.IPV4)
+				releaseIP(c, ipam.Address.IPV6)
+			}
+		}()
```
We still need to release any allocated IPs on later failure for the delegated plugin case as well, right?

wedaly replied:
there's a `defer` on line 405 above to invoke the delegated plugin's DEL, which should release the IP. I was thinking we'd skip the call to cilium-daemon's `IPAMReleaseIP`, but maybe there's some cleanup there we'd still need?

joestringer replied:
Ah right, I just missed this. No, as far as I know this should be good enough.

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -599,6 +649,18 @@ func cmdDel(args *skel.CmdArgs) error {
+		return fmt.Errorf("did not receive configuration from cilium-agent")
+	}
+
+	conf := *configResult.Status
```
Is there a reason we need this in the delete but not in the add case? I would have thought that the CNI binary still has access to the configuration object to be able to detect that delegated IPAM is in use, even from a `CNI DEL` call.

wedaly replied:
yep, the ADD command already had a call to `c.ConfigGet()` to read `conf.DatapathMode`

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -599,6 +649,18 @@ func cmdDel(args *skel.CmdArgs) error {
+	// Retrieve configuration so we can tell whether delegated IPAM is enabled or not.
+	configResult, err := c.ConfigGet()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve configuration from cilium-agent: %s", err)
```
Typically we will always use the wrap formatting strings to embed other errors:
```suggestion
		return fmt.Errorf("unable to retrieve configuration from cilium-agent: %w", err)
```

wedaly replied:
updated

joestringer commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -599,6 +649,18 @@ func cmdDel(args *skel.CmdArgs) error {
+	}
+
+	if configResult == nil || configResult.Status == nil {
+		return fmt.Errorf("did not receive configuration from cilium-agent")
```
super minor nit: I think this is actually something like `received empty configuration object from cilium-agent`, otherwise if the agent simply was notable to send the configuration then it would error out in the check above..?

wedaly replied:
updated

Commit:
pkg/ipam: add IPAMDelegatedPlugin option

By specifying --ipam=delegated-plugin, users will be able
to configure Cilium to use CNI plugin delegation
for IPAM, as described in the CNI spec:
https://www.cni.dev/docs/spec/#section-4-plugin-delegation

Signed-off-by: Will Daly <widaly@microsoft.com>

Commit:
pkg/option: validate daemon configuration with IPAMDelegatedPlugin

When using IPAMDelegatedPlugin, the CNI invokes a separate
binary (the delegated plugin) to assign IPs. This means
that the Cilium daemon cannot assign itself IPs, so validate
that LocalRouter and EndpointHealthChecking are disabled
when using IPAMDelegatedPlugin.

Co-authored-by: Tobias Klauser <tobias.klauser@gmail.com>

Signed-off-by: Will Daly <widaly@microsoft.com>

Commit:
pkg/ipam: use no-op IPAM allocator for IPAMDelegatedPlugin

When using IPAMDelegatedPlugin, the CNI handles IPAM
without involving the Cilium daemon, so the daemon
should use a no-op IPAM manager. (In practice, the no-op
IPAM manager should never be invoked, since the CNI won't
ask the daemon for IPs, and validation prevents options
that require the daemon to assign itself an IP.)

Signed-off-by: Will Daly <widaly@microsoft.com>

Commit:
cilium-cni: extract helper method getConfigFromCiliumAgent

The DEL cmd will also need to retrieve daemon configuration,
so extract the ConfigGet call into a helper method.

This commit also updates the error message for
when the daemon returns an empty config.

Signed-off-by: Will Daly <widaly@microsoft.com>

Commit:
cilium-cni: invoke delegated plugin IPAM

When the IPAM mode is set to IPAMDelegatedPlugin, Cilium CNI will
invoke the delegated plugin identified by the ipam.type field in
the CNI configuration.

This follows the delegated plugin protocol defined in the CNI spec:
https://www.cni.dev/docs/spec/#delegated-plugin-protocol

Signed-off-by: Will Daly <widaly@microsoft.com>

Commit:
cilium-cni: parse IPAM.type JSON field in netconf

When using IPAMDelegatedPlugin, the IPAM.type field in
the CNI configuration tells the CNI which delegated IPAM plugin
binary to invoke. This commit updates Cilium's NetConf structure
so the Go JSON parser deserializes this field (previously it was
always empty because another field in the struct was also named
"ipam").

Signed-off-by: Will Daly <widaly@microsoft.com>

Commit:
vendor: github.com/containernetworking/cni/pkg/invoke

When using IPAMDelegatedPlugin, the CNI needs to invoke
a separate binary for IPAM (the delegated plugin).
The package github.com/containernetworking/cni/pkg/invoke
provides helper code to invoke the delegated plugin
(this is what the CNI reference plugins use)

Signed-off-by: Will Daly <widaly@microsoft.com>

gandro commented on pkg/ipam/noop_allocator.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+func (n *noOpAllocator) Dump() (map[string]string, string) {
+	return nil, ""
```
tiny nit: The second `string` is used to display the IPAM status in `cilium status`. 

For example, here it's `24/34 allocated`.

```console
$ cilium status
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.21+ (v1.21.9-eks-0d102a7) [linux/amd64]
Cilium:                 Ok   1.10.10 (v1.10.10-024ecfc)
[...]
IPAM:                   IPv4: 24/34 allocated

```

We could return a static string like `"delegated to plugin"` here, instead of not showing anything.

wedaly replied:
thanks! will update this

gandro reviewed (approved):
Looks good to me overall! I'm not super familiar with this new CNI feature, but I didn't spot any obvious bugs.

tklauser commented on pkg/option/config.go:
```diff
@@ -3381,6 +3385,24 @@ func (c *DaemonConfig) checkIPv6NativeRoutingCIDR() error {
+		}
+		if c.EnableEndpointHealthChecking {
+			return fmt.Errorf("EnableEndpointHealthChecking must be disabled with IPAMDelegatedPlugin")
+		}
```
For more actionable and user-friendly error messages, I'd suggest to change these to explicitly mention the CLI option names, not the internal names:

```suggestion
		if c.EnableIPv4 && c.LocalRouterIPv4 == "" {
			return fmt.Errorf("--%s must be provided when IPv4 is enabled with --%s=%s", LocalRouterIPv4, IPAM, IPAMDelegatedPlugin)
		}
		if c.EnableIPv6 && c.LocalRouterIPv6 == "" {
			return fmt.Errorf("--%s must be provided when IPv6 is enabled with --%s=%s", LocalRouterIPv6, IPAM, IPAMDelegatedPlugin)
		}
		if c.EnableEndpointHealthChecking {
			return fmt.Errorf("--%s must be disabled with --%s=%s", EnableEndpointHealthChecking, IPAM, IPAMDelegatedPlugin)
		}
```

tklauser commented on pkg/option/config_test.go:
```diff
@@ -609,6 +609,67 @@ func TestCheckIPv6NativeRoutingCIDR(t *testing.T) {
+			err := tt.d.checkIPAMDelegatedPlugin()
+			if !reflect.DeepEqual(tt.expectErr, err) {
+				t.Errorf("expected error %q but got %q", tt.expectErr, err)
+			}
```
Checking the actual error for deep equality is not very idiomatic in Go. I'd suggest to just check whether an error is returned or not, i.e. make `expectErr` a `bool` and then change this to something like:

```suggestion
			hasErr := err != nil
			if tt.expectErr && !hasErr {
				t.Errorf("expected error but got none")
			} else if !expectErr && hasErr {
				t.Errorf("expected no error but got %v", err)
			}
```

wedaly replied:
thanks, updated!

tklauser reviewed (commented):
Thanks and sorry for taking so long to review this.

This LGTM apart from the minor comments inline.

tklauser reviewed (approved):
Thanks @wedaly

christarazi commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -107,6 +107,19 @@ func ipv4IsEnabled(ipam *models.IPAMResponse) bool {
+func getConfigFromCiliumAgent(client *client.Client) (*models.DaemonConfigurationStatus, error) {
+	configResult, err := client.ConfigGet()
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve configuration from cilium-agent: %s", err)
```
Nit: `%w` for errors

christarazi commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -110,11 +110,11 @@ func ipv4IsEnabled(ipam *models.IPAMResponse) bool {
 	configResult, err := client.ConfigGet()
 	if err != nil {
-		return nil, fmt.Errorf("unable to retrieve configuration from cilium-agent: %s", err)
+		return nil, fmt.Errorf("unable to retrieve configuration from cilium-agent: %w", err)
```
The changes in this commit should probably be squashed with the previous

christarazi commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -120,6 +121,27 @@ func getConfigFromCiliumAgent(client *client.Client) (*models.DaemonConfiguratio
+	podName := string(cniArgs.K8S_POD_NAMESPACE) + "/" + string(cniArgs.K8S_POD_NAME)
+	ipam, err := client.IPAMAllocate("", podName, true)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to allocate IP via local cilium agent: %s", err)
```
`%w`

christarazi replied:
Still seeing `%s` :slightly_smiling_face:

christarazi replied:
@wedaly ^

wedaly replied:
thanks! sorry I missed that, updated!

christarazi commented on pkg/ipam/noop_allocator.go:
```diff
@@ -33,7 +33,7 @@ func (n *noOpAllocator) AllocateNextWithoutSyncUpstream(owner string) (*Allocati
 
 func (n *noOpAllocator) Dump() (map[string]string, string) {
-	return nil, ""
+	return nil, "delegated to plugin"
```
This commit can also be squashed with the commit that introduced the code

christarazi commented on pkg/option/config.go:
```diff
@@ -3422,13 +3422,13 @@ func (c *DaemonConfig) checkIPAMDelegatedPlugin() error {
 		// must be disabled.
 		if c.EnableIPv4 && c.LocalRouterIPv4 == "" {
-			return fmt.Errorf("LocalRouterIPv4 must be provided when IPv4 is enabled with IPAMDelegatedPlugin")
+			return fmt.Errorf("--%s must be provided when IPv4 is enabled with --%s=%s", LocalRouterIPv4, IPAM, ipamOption.IPAMDelegatedPlugin)
```
And same with this commit re: squashing

christarazi reviewed (changes requested):
Changes LGTM overall, noted a few minor things to address. I also noticed that most of the commits don't contain a body describing the "why". Would you mind adding some context into the commits, that way we have the full context in the git history?

christarazi reviewed (approved):
LGTM, thanks for adding context to the commits. Please add the copyright header (can be found from other existing files) to the newly added files.
